File: kits/localytics-kit/src/main/java/com/mparticle/kits/LocalyticsKit.java
package com.mparticle.kits;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.os.Bundle;
import android.text.TextUtils;

import com.localytics.android.Localytics;
import com.localytics.android.LocalyticsActivityLifecycleCallbacks;
import com.localytics.android.PushReceiver;
import com.localytics.android.ReferralReceiver;
import com.mparticle.MPEvent;
import com.mparticle.MParticle;
import com.mparticle.commerce.CommerceEvent;
import com.mparticle.commerce.Product;
import com.mparticle.internal.CommerceEventUtil;
import com.mparticle.internal.ConfigManager;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class LocalyticsKit extends KitIntegration implements KitIntegration.EventListener, KitIntegration.CommerceListener, KitIntegration.AttributeListener, KitIntegration.PushListener {
    private static final String API_KEY = "appKey";
    private static final String CUSTOM_DIMENSIONS = "customDimensions";
    private static final String RAW_LTV = "trackClvAsRawValue";
    private JSONArray customDimensionJson = null;
    private boolean trackAsRawLtv = false;
    private LocalyticsActivityLifecycleCallbacks callbacks;

    @Override
    public String getName() {
        return "Localytics";
    }

    @Override
    protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
        if (callbacks == null) {
            callbacks = new LocalyticsActivityLifecycleCallbacks(getContext(), getSettings().get(API_KEY));
        }

        try {
            customDimensionJson = new JSONArray(properties.get(CUSTOM_DIMENSIONS));
        } catch (Exception jse) {
        }
        trackAsRawLtv = Boolean.parseBoolean(getSettings().get(RAW_LTV));
        return null;
    }

    @Override
    public void setLocation(Location location) {
        Localytics.setLocation(location);
    }

    @Override
    public void setInstallReferrer(Intent intent) {
        new ReferralReceiver().onReceive(getContext(), intent);
    }

    @Override
    public void setUserIdentity(MParticle.IdentityType identityType, String id) {
        if (identityType.equals(MParticle.IdentityType.Email)) {
            Localytics.setCustomerEmail(id);
        } else if (identityType.equals(MParticle.IdentityType.CustomerId)) {
            Localytics.setCustomerId(id);
        } else {
            Localytics.setIdentifier(identityType.name(), id);
        }
    }

    @Override
    public void removeUserIdentity(MParticle.IdentityType identityType) {
        if (identityType.equals(MParticle.IdentityType.Email)) {
            Localytics.setCustomerEmail("");
        } else if (identityType.equals(MParticle.IdentityType.CustomerId)) {
            Localytics.setCustomerId("");
        } else {
            Localytics.setIdentifier(identityType.name(), "");
        }
    }

    @Override
    public List<ReportingMessage> logout() {
        return null;
    }

    @Override
    public void setUserAttribute(String key, String value) {
        HashSet<String> attributeKeys = null;
        if (customDimensionJson != null) {
            try {
                for (int i = 0; i < customDimensionJson.length(); i++) {
                    JSONObject dimension = customDimensionJson.getJSONObject(i);
                    if (dimension.getString("maptype").equals("UserAttributeClass.Name")) {
                        String attributeName = dimension.getString("map");
                        if (!TextUtils.isEmpty(value)) {
                            if (attributeKeys == null) {
                                attributeKeys = new HashSet<String>();
                            }
                            attributeKeys.add(attributeName);
                            int dimensionIndex = Integer.parseInt(dimension.getString("value").substring("Dimension ".length()));
                            Localytics.setCustomDimension(dimensionIndex, value);
                        }
                    }
                }
            } catch (Exception e) {
                ConfigManager.log(MParticle.LogLevel.DEBUG, "Exception while mapping mParticle user attributes to Localytics custom dimensions: " + e.toString());
            }
        }

        if (attributeKeys == null || !attributeKeys.contains(key)) {
            if (key.equalsIgnoreCase(MParticle.UserAttributes.FIRSTNAME)) {
                Localytics.setCustomerFirstName(value);
            } else if (key.equalsIgnoreCase(MParticle.UserAttributes.LASTNAME)) {
                Localytics.setCustomerLastName(value);
            } else {
                Localytics.setProfileAttribute(KitUtils.sanitizeAttributeKey(key), value);
            }
        }
    }

    @Override
    public void removeUserAttribute(String key) {

    }

    @Override
    public List<ReportingMessage> setOptOut(boolean optOutStatus) {
        Localytics.setOptedOut(optOutStatus);
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(new ReportingMessage(this, ReportingMessage.MessageType.OPT_OUT, System.currentTimeMillis(), null));
        return messageList;
    }

    @Override
    public List<ReportingMessage> leaveBreadcrumb(String breadcrumb) {
        return null;
    }

    @Override
    public List<ReportingMessage> logError(String message, Map<String, String> errorAttributes) {
        return null;
    }

    @Override
    public List<ReportingMessage> logException(Exception exception, Map<String, String> exceptionAttributes, String message) {
        return null;
    }

    @Override
    public List<ReportingMessage> logEvent(MPEvent event) {
        Localytics.tagEvent(event.getEventName(), event.getInfo());
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(ReportingMessage.fromEvent(this, event));
        return messageList;
    }

    @Override
    public List<ReportingMessage> logScreen(String screenName, Map<String, String> eventAttributes) {
        Localytics.tagScreen(screenName);
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(
                new ReportingMessage(this, ReportingMessage.MessageType.SCREEN_VIEW, System.currentTimeMillis(), eventAttributes)
                        .setScreenName(screenName)
        );
        return messageList;
    }

    @Override
    public List<ReportingMessage> logLtvIncrease(BigDecimal valueIncreased, BigDecimal totalValue, String eventName, Map<String, String> contextInfo) {
        int multiplier = trackAsRawLtv ? 1 : 100;
        Localytics.tagEvent(eventName, contextInfo, (long) valueIncreased.doubleValue() * multiplier);
        List<ReportingMessage> messageList = new LinkedList<ReportingMessage>();
        messageList.add(ReportingMessage.fromEvent(this, new MPEvent.Builder(eventName, MParticle.EventType.Transaction).info(contextInfo).build()));
        return messageList;
    }

    @Override
    public List<ReportingMessage> logEvent(CommerceEvent event) {
        List<ReportingMessage> messages = new LinkedList<ReportingMessage>();
        if (!TextUtils.isEmpty(event.getProductAction()) && (
                event.getProductAction().equalsIgnoreCase(Product.PURCHASE)) ||
                event.getProductAction().equalsIgnoreCase(Product.REFUND)) {
            Map<String, String> eventAttributes = new HashMap<String, String>();
            CommerceEventUtil.extractActionAttributes(event, eventAttributes);
            int multiplier = trackAsRawLtv ? 1 : 100;
            if (event.getProductAction().equalsIgnoreCase(Product.REFUND)) {
                multiplier *= -1;
            }
            double total = event.getTransactionAttributes().getRevenue() * multiplier;
            Localytics.tagEvent(String.format("eCommerce - %s", event.getProductAction(), eventAttributes, (long) total));
            messages.add(ReportingMessage.fromEvent(this, event));
            return messages;
        }
        List<MPEvent> eventList = CommerceEventUtil.expand(event);
        if (eventList != null) {
            for (int i = 0; i < eventList.size(); i++) {
                try {
                    logEvent(eventList.get(i));
                    messages.add(ReportingMessage.fromEvent(this, event));
                } catch (Exception e) {
                    ConfigManager.log(MParticle.LogLevel.WARNING, "Failed to call tagEvent to Localytics kit: " + e.toString());
                }
            }
        }
        return messages;
    }

    @Override
    public boolean willHandlePushMessage(Set<String> keyset) {
        return keyset.contains("ll");
    }

    @Override
    public void onPushMessageReceived(Context context, Intent extras) {
        new PushReceiver().onReceive(context, extras);
    }

    @Override
    public boolean onPushRegistration(String token, String senderId) {
        Localytics.setPushRegistrationId(token);
        return true;
    }
}

File: kits/localytics/localytics-6/src/main/kotlin/com/mparticle/kits/LocalyticsKit.kt
package com.mparticle.kits

import android.app.Application
import android.content.Context
import android.content.Intent
import android.location.Location
import com.localytics.androidx.CallToActionListener
import com.localytics.androidx.Campaign
import com.localytics.androidx.FirebaseService
import com.localytics.androidx.Localytics
import com.localytics.androidx.ReferralReceiver
import com.mparticle.MPEvent
import com.mparticle.MParticle
import com.mparticle.MParticle.IdentityType
import com.mparticle.commerce.CommerceEvent
import com.mparticle.commerce.Product
import com.mparticle.internal.Logger
import com.mparticle.internal.MPUtility
import com.mparticle.kits.KitIntegration.AttributeListener
import com.mparticle.kits.KitIntegration.CommerceListener
import com.mparticle.kits.KitIntegration.EventListener
import com.mparticle.kits.KitIntegration.LogoutListener
import com.mparticle.kits.KitIntegration.PushListener
import com.mparticle.kits.ReportingMessage
import org.json.JSONArray
import java.math.BigDecimal
import java.util.HashMap
import java.util.LinkedList

class LocalyticsKit :
    KitIntegration(),
    KitIntegration.EventListener,
    CommerceListener,
    AttributeListener,
    LogoutListener,
    PushListener,
    CallToActionListener {
    private var customDimensionJson: JSONArray? = null
    private var trackAsRawLtv = false

    override fun getName(): String = KIT_NAME

    public override fun onKitCreate(
        settings: Map<String, String>,
        context: Context,
    ): List<ReportingMessage> {
        try {
            customDimensionJson = JSONArray(settings[CUSTOM_DIMENSIONS])
        } catch (jse: Exception) {
        }
        trackAsRawLtv = java.lang.Boolean.parseBoolean(settings[RAW_LTV])
        Localytics.setOption("ll_app_key", settings[API_KEY])
        Localytics.autoIntegrate((context.applicationContext as Application))

        // after a reset() call, we need to set Provivacy OptedOut back to false, so the kit can run normally
        Localytics.setCallToActionListener(this)
        Localytics.setLoggingEnabled(MParticle.getInstance()!!.environment == MParticle.Environment.Development)
        return emptyList()
    }

    override fun setLocation(location: Location) {
        Localytics.setLocation(location)
    }

    override fun setInstallReferrer(intent: Intent) {
        ReferralReceiver().onReceive(context, intent)
    }

    override fun setUserIdentity(
        identityType: IdentityType,
        id: String,
    ) {
        when (identityType) {
            IdentityType.Email -> {
                Localytics.setCustomerEmail(id)
            }
            IdentityType.CustomerId -> {
                Localytics.setCustomerId(id)
            }
            else -> {
                Localytics.setIdentifier(identityType.name, id)
            }
        }
    }

    override fun removeUserIdentity(identityType: IdentityType) {
        when (identityType) {
            IdentityType.Email -> {
                Localytics.setCustomerEmail("")
            }
            IdentityType.CustomerId -> {
                Localytics.setCustomerId("")
            }
            else -> {
                Localytics.setIdentifier(identityType.name, "")
            }
        }
    }

    override fun logout(): List<ReportingMessage> = emptyList()

    fun getDimensionIndexForAttribute(key: String?): Int {
        if (customDimensionJson == null) {
            try {
                customDimensionJson = JSONArray(settings[CUSTOM_DIMENSIONS])
            } catch (jse: Exception) {
            }
        }
        customDimensionJson?.let { customDimensionJson ->
            try {
                for (i in 0 until customDimensionJson.length()) {
                    val dimension = customDimensionJson.getJSONObject(i)
                    if (dimension.getString("maptype") == "UserAttributeClass.Name") {
                        val attributeName = dimension.getString("map")
                        if (key.equals(attributeName, true)) {
                            return dimension
                                .getString("value")
                                .substring("Dimension ".length)
                                .toInt()
                        }
                    }
                }
            } catch (e: Exception) {
                Logger.debug("Exception while mapping mParticle user attributes to Localytics custom dimensions: $e")
            }
        }
        return -1
    }

    override fun setUserAttribute(
        key: String,
        value: String,
    ) {
        val dimensionIndex = getDimensionIndexForAttribute(key)
        if (dimensionIndex >= 0) {
            Localytics.setCustomDimension(dimensionIndex, value)
        }
        when {
            key.equals(MParticle.UserAttributes.FIRSTNAME, true) -> {
                Localytics.setCustomerFirstName(value)
            }
            key.equals(MParticle.UserAttributes.LASTNAME, true) -> {
                Localytics.setCustomerLastName(value)
            }
            else -> {
                Localytics.setProfileAttribute(KitUtils.sanitizeAttributeKey(key), value)
            }
        }
    }

    override fun setUserAttributeList(
        key: String,
        list: List<String>,
    ) {
        val array = list.toTypedArray()
        Localytics.setProfileAttribute(key, array)
    }

    override fun supportsAttributeLists(): Boolean = true

    override fun setAllUserAttributes(
        attributes: Map<String, String>,
        attributeLists: Map<String, List<String>>,
    ) {
        for ((key, value) in attributes) {
            setUserAttribute(key, value)
        }
        for ((key, value) in attributeLists) {
            setUserAttributeList(key, value)
        }
    }

    override fun removeUserAttribute(key: String) {
        Localytics.deleteProfileAttribute(key)
    }

    override fun setOptOut(optOutStatus: Boolean): List<ReportingMessage> {
        Localytics.setOptedOut(optOutStatus)
        val messageList: MutableList<ReportingMessage> = LinkedList()
        messageList.add(
            ReportingMessage(
                this,
                ReportingMessage.MessageType.OPT_OUT,
                System.currentTimeMillis(),
                null,
            ),
        )
        return messageList
    }

    override fun leaveBreadcrumb(breadcrumb: String): List<ReportingMessage> = emptyList()

    override fun logError(
        message: String,
        errorAttributes: Map<String, String>,
    ): List<ReportingMessage> = emptyList()

    override fun logException(
        exception: Exception,
        exceptionAttributes: Map<String, String>,
        message: String,
    ): List<ReportingMessage> = emptyList()

    override fun logEvent(event: MPEvent): List<ReportingMessage> {
        val duration = event.length
        var info = event.customAttributeStrings
        if (duration != null) {
            if (info == null) {
                info = HashMap()
            }
            info["event_duration"] = duration.toString()
        }
        Localytics.tagEvent(event.eventName, info)
        val messageList: MutableList<ReportingMessage> = LinkedList()
        messageList.add(ReportingMessage.fromEvent(this, event))
        return messageList
    }

    override fun logScreen(
        screenName: String,
        eventAttributes: Map<String, String>,
    ): List<ReportingMessage> {
        Localytics.tagScreen(screenName)
        val messageList: MutableList<ReportingMessage> = LinkedList()
        messageList.add(
            ReportingMessage(
                this,
                ReportingMessage.MessageType.SCREEN_VIEW,
                System.currentTimeMillis(),
                eventAttributes,
            ).setScreenName(screenName),
        )
        return messageList
    }

    override fun logLtvIncrease(
        valueIncreased: BigDecimal,
        totalValue: BigDecimal,
        eventName: String,
        contextInfo: Map<String, String>,
    ): List<ReportingMessage> {
        val multiplier = if (trackAsRawLtv) 1 else 100
        Localytics.tagEvent(eventName, contextInfo, valueIncreased.toDouble().toLong() * multiplier)
        val messageList: MutableList<ReportingMessage> = LinkedList()
        messageList.add(
            ReportingMessage.fromEvent(
                this,
                MPEvent
                    .Builder(eventName, MParticle.EventType.Transaction)
                    .customAttributes(contextInfo)
                    .build(),
            ),
        )
        return messageList
    }

    override fun logEvent(event: CommerceEvent): List<ReportingMessage> {
        val messages: MutableList<ReportingMessage> = LinkedList()
        if (!KitUtils.isEmpty(event.productAction) &&
            event.productAction.equals(
                Product.PURCHASE,
                true,
            ) ||
            event.productAction.equals(Product.REFUND, true)
        ) {
            val eventAttributes: Map<String, String> = HashMap()
            CommerceEventUtils.extractActionAttributes(event, eventAttributes)
            var multiplier = if (trackAsRawLtv) 1 else 100
            if (event.productAction.equals(Product.REFUND, true)) {
                multiplier *= -1
            }
            val total = event.transactionAttributes?.revenue?.times(multiplier)
            if (total != null) {
                Localytics.tagEvent(
                    String.format("eCommerce - %s", event.productAction),
                    eventAttributes,
                    total.toLong(),
                )
            }
            messages.add(ReportingMessage.fromEvent(this, event))
            return messages
        }
        val eventList = CommerceEventUtils.expand(event)
        if (eventList != null) {
            for (i in eventList.indices) {
                try {
                    logEvent(eventList[i])
                    messages.add(ReportingMessage.fromEvent(this, event))
                } catch (e: Exception) {
                    Logger.warning("Failed to call tagEvent to Localytics kit: $e")
                }
            }
        }
        return messages
    }

    override fun willHandlePushMessage(intent: Intent): Boolean =
        (
            intent.extras?.containsKey("ll") == true ||
                intent.extras?.containsKey("localyticsUninstallTrackingPush") == true
            ) &&
            MPUtility.isFirebaseAvailable()

    override fun onPushMessageReceived(
        context: Context,
        extras: Intent,
    ) {
        val service: Intent
        if (MPUtility.isFirebaseAvailable()) {
            service = Intent(context, FirebaseService::class.java)
            service.action = "com.google.firebase.MESSAGING_EVENT"
            service.putExtras(extras)
            context.startService(service)
        }
    }

    override fun onPushRegistration(
        token: String,
        senderId: String,
    ): Boolean {
        Localytics.setPushRegistrationId(token)
        return true
    }

    public override fun reset() {
        Localytics.setPrivacyOptedOut(true)
    }

    override fun localyticsShouldDeeplink(
        s: String,
        campaign: Campaign,
    ): Boolean = false

    override fun localyticsDidOptOut(
        b: Boolean,
        campaign: Campaign,
    ) {}

    override fun localyticsDidPrivacyOptOut(
        optedOut: Boolean,
        campaign: Campaign,
    ) {
        if (optedOut && !kitManager.isOptedOut) {
            Localytics.setPrivacyOptedOut(false)
        }
    }

    override fun localyticsShouldPromptForLocationPermissions(campaign: Campaign): Boolean = false

    override fun localyticsShouldPromptForNotificationPermissions(campaign: Campaign): Boolean = false

    companion object {
        const val API_KEY = "appKey"
        const val CUSTOM_DIMENSIONS = "customDimensions"
        private const val RAW_LTV = "trackClvAsRawValue"
        private const val KIT_NAME = "Localytics"
    }
}
